presto-main/src/main/java/com/facebook/presto/operator/MapBlockEncodingBuffers.java

```diff
@@ -0,0 +1,371 @@
+
+        keyBuffers.setupDecodedBlockAndMapPositions(decodedBlockNode.getChildren().get(0));
+        valueBuffers.setupDecodedBlockAndMapPositions(decodedBlockNode.getChildren().get(1));
+    }
+
+    @Override
+    public void accumulateRowSizes(int[] rowSizes)
+    {
+        for (int i = 0; i < positionCount; i++) {
+            rowSizes[i] += POSITION_SIZE;
+        }
+
+        ensureOffsetsOffsetsCapacity(positionCount);
+        System.arraycopy(offsets, 0, keyOffsetsOffsets, 0, positionCount + 1);
+        System.arraycopy(offsets, 0, valueOffsetsOffsets, 0, positionCount + 1);
+
+        keyBuffers.accumulateRowSizes(keyOffsetsOffsets, positionCount, rowSizes);
+        valueBuffers.accumulateRowSizes(valueOffsetsOffsets, positionCount, rowSizes);
+    }
+
+    @Override
+    protected void accumulateRowSizes(int[] offsetsOffsets, int positionCount, int[] rowSizes)
+    {
+        // The nested level positionCount could be 0.
+        if (this.positionCount == 0) {
+            return;
+        }
+
+        int lastOffset = offsetsOffsets[0];
+        for (int i = 0; i < positionCount; i++) {
+            int offset = offsetsOffsets[i + 1];
+            rowSizes[i] += POSITION_SIZE * (offset - lastOffset);
+            lastOffset = offset;
+            offsetsOffsets[i + 1] = offsets[offset];
+        }
+
+        // offsetsOffsets might be modified by the next level. Save it for the valueBuffers first.
+        ensureOffsetsOffsetsCapacity(positionCount);
+        System.arraycopy(offsetsOffsets, 0, valueOffsetsOffsets, 0, positionCount + 1);
+
+        keyBuffers.accumulateRowSizes(offsetsOffsets, positionCount, rowSizes);
+        valueBuffers.accumulateRowSizes(offsetsOffsets, positionCount, rowSizes);
+    }
+
+    @Override
+    public void setNextBatch(int positionsOffset, int batchSize)
+    {
+        this.positionsOffset = positionsOffset;
+        this.batchSize = batchSize;
+
+        if (this.positionCount == 0) {
+            return;
+        }
+
+        int beginOffset = offsets[positionsOffset];
+        int endOffset = offsets[positionsOffset + batchSize];
+
+        keyBuffers.setNextBatch(beginOffset, endOffset - beginOffset);
+        valueBuffers.setNextBatch(beginOffset, endOffset - beginOffset);
+    }
+
+    @Override
+    public void copyValues()
+    {
+        if (batchSize == 0) {
+            return;
+        }
+
+        appendNulls();
+        appendOffsets();
+        appendHashTables();
+
+        keyBuffers.copyValues();
+        valueBuffers.copyValues();
+
+        bufferedPositionCount += batchSize;
+    }
+
+    public void serializeTo(SliceOutput output)
+    {
+        writeLengthPrefixedString(output, NAME);
+
+        // Map type
+        TypeSerde.writeType(output, columnarMap.getKeyType());
+
+        keyBuffers.serializeTo(output);
+        valueBuffers.serializeTo(output);
+
+        // Hash tables
+        int hashTableSize = lastOffset * HASH_MULTIPLIER;
```

Comment:
this variable is used only in `else` branch below; let's move it there and, perhaps, inline.